### PR TITLE
Keras best metric works with resume, some refactoring and BUG fixes

### PR DESCRIPTION
**Pull request by vanpelt**
Made restoring the best metric for keras work, added a test, fixed some bugs and refactored sender.  I decided to put all of the update user process state from resume in one place.  This means we issue a summary update request on resume but it seems like a small price to pay for keeping the code unified.

Todo:

- [x] confirm run.resumed works in the auto case
- [ ] more tests

### CODE REVIEW

File: wandb/sdk/wandb_config.py [hunk 1]
```diff
@@ -113,7 +113,7 @@ def update_locked(self, d, user=None):
     def _sanitize_dict(self, config_dict, allow_val_change=None):
         sanitized = {}
         for k, v in six.iteritems(config_dict):
-            k, v = self._sanitize(k, v)
+            k, v = self._sanitize(k, v, allow_val_change)
```
Comment (vanpelt) [inline wandb/sdk/wandb_config.py:116]: Beast 🐞

File: wandb/sdk/wandb_run.py [hunk 2]
```diff
@@ -210,6 +210,12 @@ def restore(
     def use_artifact(self, artifact_or_name, type=None, aliases=None):
         pass
 
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        return exc_type is None
+
```
Comment (vanpelt) [inline wandb/sdk/wandb_run.py:218]: Noticed our dummy run wasn't a context

File: wandb/sdk/wandb_run.py [hunk 3]
```diff
@@ -559,7 +565,20 @@ def _set_teardown_hooks(self, hooks):
...
-        # TODO: Update run summary when resuming?
+        # Grab the config from resuming
+        if run_obj.HasField("config"):
+            c_dict = config_util.dict_no_value_from_proto_list(run_obj.config.update)
+            # TODO: Windows throws a wild error when this is set...
+            if "_wandb" in c_dict:
+                del c_dict["_wandb"]
+            # We update the config object here without triggering the callback
+            self.config._update(c_dict, allow_val_change=True)
+        # Update the summary, this will trigger an un-needed graphql request :(
```
Comment (adrnswanberg) [inline wandb/sdk/wandb_run.py:576]: I'm seeing a good argument for an optional `skip_update` arg in Summary::_update :)

**Review by adrnswanberg — APPROVED** [review]

File: wandb/sdk/wandb_config.py (hunk 1 above)
Comment (raubitsj) [reply wandb/sdk/wandb_config.py:116]: this was a filed issue, waiting for someone to get to it and add a test.

File: wandb/sdk/wandb_run.py (hunk 2 above)
Comment (raubitsj) [reply wandb/sdk/wandb_run.py:218]: nice

**Pull request merged**